sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.autoconfigure;
+
+import io.opentelemetry.api.internal.GuardedBy;
+import io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider;
+import io.opentelemetry.sdk.resources.Resource;
+import java.util.HashSet;
+import java.util.ServiceLoader;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/** Auto-configuration for the OpenTelemetry {@link Resource}. */
+public final class OpenTelemetryResourceAutoConfiguration {
+
+  @Nullable private static volatile Resource resource = null;
```

Comment:
I wonder...how important is it that we keep a copy of this, making a global of the configured Resource? What would be the downside if we just always re-initialized it?

Reply:
Hmm, there'd be a risk of using different `ConfigProperties` for each call. For example, initializing the SDK with a custom `ConfigProperties` implementation and then calling `getResource()` (that uses default config) could result in different resources. 
So, if somebody used the javaagent's `ConfigPropertySource` SPI to set some resource attributes they'd always be visible in the auto-configured SDK, but if you called `OpenTelemetryResource...#getResource()` without passing the  same `ConfigProperties` you'd get a different resource (constructed just from system properties and env vars).

I don't know it it's a problem though, if we document that behavior it should probably be fine. In case of the instrumentation javaagent the config's globally available, so it'd be fine to just ensure we pass it every time we call `getResource()`.

Reply:
I think if we never try to claim that there's a single global Resource, (maybe "initialize..." isn't the right name for the method), that this would be fine. It's just a factory for Resource instances based on config.

@anuraaga any thoughts on this?

Reply:
Done - PTAL.

Reply:
I think this looks good. Would like @anuraaga to chime in before merging.